### PR TITLE
swiftlint 0.16.0

### PR DESCRIPTION
**Pull request by jpsim**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by jpsim** [comment]
I can reproduce the `brew test` failures locally by running the same steps as are run on Jenkins, but I don't yet know the cause of the failures.

How would you recommend I troubleshoot this? In `CONTRIBUTING.md`, I see there's a `brew install --build-from-source foo` command, but that will still use the commit specified in the formula in this PR. Is there a way to specify the location for the source?

Installing and running the test outside of Homebrew's environment passes the test, so I have to troubleshoot within Homebrew's environment.

**Comment by jpsim** [comment]
Sharing my process for others to learn from:

I couldn't figure out a good way to test a local git repo as the source for a homebrew formula (the `head "file://..."` trick just seems to work with tarballs, not entire source trees or git repos.

So I ended up using `brew edit` and a temporary remote git branch:

```
$ brew edit swiftlint
$ # change head to
$ # head "https://github.com/realm/SwiftLint.git", :branch => "homebrew-0.16.0-troubleshoot-please-ignore"
$ brew install swiftlint --HEAD
$ brew test swiftlint --verbose
```

So now to troubleshoot, I can push attempts to debug to this branch and repeat the process. Super inelegant, but the best I could think of without sinking too much time into the process of the fix, focusing more on the actual fix.

**Comment by jpsim** [comment]
Running with `brew test swiftlint --verbose --no-sandbox` works fine, so the culprit here must be the same as Homebrew/legacy-homebrew#50211. This makes sense since SwiftLint is _based_ on SourceKitten.

I could dumb down this test like we did with SourceKitten (just run `sourcekitten version`, which doesn't make an XPC connection with Xcode).

Do the Homebrew folks here have any other suggestions or preferences?

**Comment by jpsim** [comment]
@ilovezfs thoughts?

**Comment by MikeMcQuaid** [comment]
> I couldn't figure out a good way to test a local git repo as the source for a homebrew formula (the head "file://..." trick just seems to work with tarballs, not entire source trees or git repos.

Another option may be to install the default repo with `brew install --git --interactive` which will allow you to modify files before installation.

**Comment by jpsim** [comment]
Thanks for the tip @MikeMcQuaid! How about this PR?

**Comment by ilovezfs** [comment]
This PR is :-1: until the sandboxing problem is reported upstream. It doesn't affect the current version, only the new one.

**Review by ilovezfs — CHANGES_REQUESTED** [review]
lint shouldn't have to violate the sandbox to run. Let's get this fixed upstream since it's a regression from 0.15.0

File: Formula/swiftlint.rb [hunk 1]
```diff
@@ -19,7 +19,8 @@ def install
   end
 
   test do
-    (testpath/"Test.swift").write "import Foundation\n"
```
Comment (ilovezfs) [inline Formula/swiftlint.rb]: please put this back the way it was

**Comment by jpsim** [comment]
This is not a bug upstream. I'm actually very surprised that this issue hasn't manifested itself before now given that SwiftLint _has always_ made an XPC connection to SourceKit.

**Comment by ilovezfs** [comment]
The exact same test passes with the old version.

**Comment by jpsim** [comment]
I'm aware, but there's no architectural change that would account for the difference. But that's besides the point. It's expected behavior for SwiftLint to make an XPC connection, which the sandbox mechanism registers as an unauthorized  file system write but in reality is absolutely harmless.

**Comment by ilovezfs** [comment]
You're assuming the cause is the xpc connection, which it must not be if the prior version also uses that same mechanism. Running the executable in the sandbox actually emits an illegal instruction error. I recommend git bisect.

**Comment by jpsim** [comment]
Actually not assuming anything. Please see https://github.com/Homebrew/legacy-homebrew/pull/50211 for more information confirming the XPC connection is the culprit.

Notably, the "specified service did not ship in the requestor's bundle" error is what's triggering the sandbox violation. This happens even if the `sourcekitd.framework` is in the test directory:

```
com.apple.xpc.launchd[1] (com.apple.xpc.launchd.domain.pid.swiftlint.93932): Path not allowed in target domain: type = pid, path = /Applications/Xcode-8.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/sourcekitd.framework/Versions/A/XPCServices/SourceKitService.xpc error = 147: The specified service did not ship in the requestor's bundle, origin = (unknown)
com.apple.xpc.launchd[1] (com.apple.xpc.launchd.domain.pid.swiftlint.94345): Path not allowed in target domain: type = pid, path = /private/tmp/swiftlint-test-20170119-94332-1rhz39c/usr/lib/sourcekitd.framework/Versions/A/XPCServices/SourceKitService.xpc error = 147: The specified service did not ship in the requestor's bundle, origin = (unknown)
```

**Comment by MikeMcQuaid** [comment]
> I'm aware, but there's no architectural change that would account for the difference.

> which it must not be if the prior version also uses that same mechanism. Running the executable in the sandbox actually emits an illegal instruction error.

There two comments don't really square up. Regardless, we need to figure something out here.

**Comment by jpsim** [comment]
Alright, I did quite a bit of digging to get to the bottom of this.

I had forgotten that as of a few versions ago, SwiftLint can partially recover from SourceKit failures (such as the ones that happen when running in the Sandbox), which is why Homebrew tests for this formula have _always_ had the "sourcekitd failed" error message, but still succeeded.

What's changed is that a call to SourceKit was added between 0.15.0 and 0.16.0 that doesn't handle the error case of SourceKit not responding, which happens when running in a sandboxed environment since the XPC service doesn't ship with the app's bundle. I've pushed a fix and will be releasing 0.16.1 with it shortly, at which point I'll make another Homebrew PR, including more comprehensive tests that validate the command output.

Thanks for working with me on this @ilovezfs and @MikeMcQuaid.

**Comment by ilovezfs** [comment]
Nice work @jpsim! Thanks for taking the extra time to get this sandbox friendly again and diagnose the situation.

**Comment by jpsim** [comment]
0.16.1 PR in #9110.

**Pull request closed without merging**